### PR TITLE
[Bugfix] logprobs is not compatible with the OpenAI spec #4795

### LINKED ISSUE #4795: [Bug]: `logprobs` is not compatible with the OpenAI spec
### Your current environment

I'm using Runpod Serverless vLLM (https://github.com/runpod-workers/worker-vllm) so I can't run this command. However, I confirmed that the issue is in the codebase in `main`:

https://github.com/vllm-project/vllm/blob/0fca3cdcf265cd375bca684d951702b6b7adf65a/vllm/entrypoints/openai/protocol.py

### 🐛 Describe the bug

The behavior of `logprobs=True` does not match OpenAI's.

I identified two issues:

**(1) vLLM throws an error when `logprobs=True` and `top_logprobs` is missing.**

OpenAI works fine:

```py
completion = openai_client.chat.completions.create(
  model="gpt-4-turbo-preview",
  messages=[
    {"role": "user", "content": "Hi!"}
  ],
  logprobs=True,
)
```

```
ChatCompletion(id='chatcmpl-9OY4XFK8suJ7ed0yw5vglbTsOZUt1', choices=[Choice(finish_reason='stop', index=0, logprobs=ChoiceLogprobs(content=[ChatCompletionTokenLogprob(token='Hello', bytes=[72, 101, 108, 108, 111], logprob=-0.0008963357, top_logprobs=[]), ChatCompletionTokenLogprob(token='!', bytes=[33], logprob=-9.729906e-06, top_logprobs=[]), ChatCompletionTokenLogprob(token=' How', bytes=[32, 72, 111, 119], logprob=-1.4140442e-05, top_logprobs=[]), ChatCompletionTokenLogprob(token=' can', bytes=[32, 99, 97, 110], logprob=-0.0004804817, top_logprobs=[]), ChatCompletionTokenLogprob(token=' I', bytes=[32, 73], logprob=-1.11603495e-05, top_logprobs=[]), ChatCompletionTokenLogprob(token=' assist', bytes=[32, 97, 115, 115, 105, 115, 116], logprob=-0.30324343, top_logprobs=[]), ChatCompletionTokenLogprob(token=' you', bytes=[32, 121, 111, 117], logprob=-5.5122365e-07, top_logprobs=[]), ChatCompletionTokenLogprob(token=' today', bytes=[32, 116, 111, 100, 97, 121], logprob=-1.700133e-05, top_logprobs=[]), ChatCompletionTokenLogprob(token='?', bytes=[63], logprob=-0.001247851, top_logprobs=[])]), message=ChatCompletionMessage(content='Hello! How can I assist you today?', role='assistant', function_call=None, tool_calls=None))], created=1715637873, model='gpt-4-0125-preview', object='chat.completion', system_fingerprint=None, usage=CompletionUsage(completion_tokens=9, prompt_tokens=9, total_tokens=18))
```

vLLM breaks:

```python
completion = vllm_client.chat.completions.create(
  model="...my-llama3-8b-finetune...",
  messages=[
    {"role": "user", "content": "Hi!"}
  ],
  logprobs=True,
)
```

```
ChatCompletion(id=None, choices=None, created=None, model=None, object='error', system_fingerprint=None, usage=None, code=400, message='Top logprobs must be set when logprobs is.', param=None, type='BadRequestError')
```

via https://github.com/vllm-project/vllm/blob/0fca3cdcf265cd375bca684d951702b6b7adf65a/vllm/entrypoints/openai/protocol.py#L162

**(2) Even wtih `top_logprobs=1`, the behavior doesn't match.**

OpenAI:

```python
completion = openai_client.chat.completions.create(
  model="gpt-4-turbo-preview",
  messages=[
    {"role": "user", "content": "Hi!"}
  ],
  logprobs=True,
  top_logprobs=1,
)
```

```
ChatCompletion(id='chatcmpl-9OY4PwigZVtoM6vHXELby3NWqCyaX', choices=[Choice(finish_reason='stop', index=0, logprobs=ChoiceLogprobs(content=[ChatCompletionTokenLogprob(token='Hello', bytes=[72, 101, 108, 108, 111], logprob=-0.0008963357, top_logprobs=[TopLogprob(token='Hello', bytes=[72, 101, 108, 108, 111], logprob=-0.0008963357)]), ChatCompletionTokenLogprob(token='!', bytes=[33], logprob=-9.729906e-06, top_logprobs=[TopLogprob(token='!', bytes=[33], logprob=-9.729906e-06)]), ChatCompletionTokenLogprob(token=' How', bytes=[32, 72, 111, 119], logprob=-1.4140442e-05, top_logprobs=[TopLogprob(token=' How', bytes=[32, 72, 111, 119], logprob=-1.4140442e-05)]), ChatCompletionTokenLogprob(token=' can', bytes=[32, 99, 97, 110], logprob=-0.0004804817, top_logprobs=[TopLogprob(token=' can', bytes=[32, 99, 97, 110], logprob=-0.0004804817)]), ChatCompletionTokenLogprob(token=' I', bytes=[32, 73], logprob=-1.11603495e-05, top_logprobs=[TopLogprob(token=' I', bytes=[32, 73], logprob=-1.11603495e-05)]), ChatCompletionTokenLogprob(token=' assist', bytes=[32, 97, 115, 115, 105, 115, 116], logprob=-0.3164038, top_logprobs=[TopLogprob(token=' assist', bytes=[32, 97, 115, 115, 105, 115, 116], logprob=-0.3164038)]), ChatCompletionTokenLogprob(token=' you', bytes=[32, 121, 111, 117], logprob=-5.5122365e-07, top_logprobs=[TopLogprob(token=' you', bytes=[32, 121, 111, 117], logprob=-5.5122365e-07)]), ChatCompletionTokenLogprob(token=' today', bytes=[32, 116, 111, 100, 97, 121], logprob=-1.50940705e-05, top_logprobs=[TopLogprob(token=' today', bytes=[32, 116, 111, 100, 97, 121], logprob=-1.50940705e-05)]), ChatCompletionTokenLogprob(token='?', bytes=[63], logprob=-0.0011334282, top_logprobs=[TopLogprob(token='?', bytes=[63], logprob=-0.0011334282)])]), message=ChatCompletionMessage(content='Hello! How can I assist you today?', role='assistant', function_call=None, tool_calls=None))], created=1715637865, model='gpt-4-0125-preview', object='chat.completion', system_fingerprint=None, usage=CompletionUsage(completion_tokens=9, prompt_tokens=9, total_tokens=18))
```

vLLM:

```python
completion = vllm_client.chat.completions.create(
  model="...my-llama3-8b-finetune...",
  messages=[
    {"role": "user", "content": "Hi!"}
  ],
  logprobs=True,
  top_logprobs=1,
)
```

```
ChatCompletion(id='cmpl-c9459bd09bb24a028fef65190d22248d', choices=[Choice(finish_reason='stop', index=0, logprobs=ChoiceLogprobs(content=None, text_offset=[0, 2, 3, 7, 9, 14, 18, 24, 25], token_logprobs=[-1.6077232360839844, -1.1920228004455566, -0.0803707167506218, -0.8764647841453552, -0.2521621584892273, -1.823885577323381e-05, -0.020511768758296967, -0.007883979007601738, -0.0015675650211051106], tokens=['Hi', '!', 'ĠHow', "'s", 'Ġyour', 'Ġday', 'Ġgoing', '?', '<|im_end|>'], top_logprobs=[{'Hey': -1.2327232360839844, 'Hi': -1.6077232360839844}, {'!': -1.1920228004455566, 'Ġthere': -0.44202280044555664}, {'ĠHow': -0.0803707167506218}, {"'s": -0.8764647841453552, 'Ġare': -0.6264647841453552}, {'Ġyour': -0.2521621584892273}, {'Ġday': -1.823885577323381e-05}, {'Ġgoing': -0.020511768758296967}, {'?': -0.007883979007601738}, {'<|im_end|>': -0.0015675650211051106}]), message=ChatCompletionMessage(content="Hi! How's your day going?", role='assistant', function_call=None, tool_calls=None))], created=2362885, model='REDACTED', object='chat.completion', system_fingerprint=None, usage=CompletionUsage(completion_tokens=9, prompt_tokens=12, total_tokens=21))
```

Notice that, for example, `token_logprobs` comes up with vLLM but not with OpenAI.

---

These issues break libraries expecting OpenAI-compatible responses, e.g. Rust's async_openai we are using.

### PR DESCRIPTION
**Pull request by Etelis**
### Description

This PR addresses several issues related to logprobs handling and sampling in the vLLM project.

- **Changed default value of `top_logprobs`**:
  - In `ChatCompletionRequest` class, changed the default value of `top_logprobs` from `None` to `0`.
  - Commented out the check in `to_sampling_params` method to allow `logprobs` being true without `top_logprobs` being set.

- **Adjusted logprobs handling**:
  - In the `chat_completion_full_generator` function, changed `num_output_top_logprobs` to `request.top_logprobs` instead of `request.logprobs` to ensure only the needed number of top logprobs are taken. This change addresses the issue where vLLM throws an error when `logprobs=True` and `top_logprobs` is missing.

- **Fixed logprob sampling range**:
  - In the `_get_sampled_logprob_if_needed` function, changed `top_ranks = range(1, num_logprobs + 1)` to `top_ranks = range(1, num_logprobs)` to prevent oversampling by 1, aligning the sampling method with OpenAI's behavior.

---

FIX [#4795](https://github.com/vllm-project/vllm/issues/4795)
**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by DarkLight1337** [comment]
> In the `_get_sampled_logprob_if_needed` function, changed `top_ranks = range(1, num_logprobs + 1)` to `top_ranks = range(1, num_logprobs)` to prevent oversampling by 1, aligning the sampling method with OpenAI's behavior.

Actuallly, since `range` doesn't include the end value, the original one is correct.

Besides, OpenAI spec indicates that up to `logprob+1` results can be returned, so I guess it should additionally include the logprob of the current token.

**Comment by DarkLight1337** [comment]
How about we split this PR into two? We can get the first two issues fixed first before working on the third issue since that one involves more of the internals.

File: vllm/entrypoints/openai/serving_chat.py [hunk 1]
```diff
@@ -367,7 +367,7 @@ async def chat_completion_full_generator(
                 logprobs = self._create_logprobs(
                     token_ids=token_ids,
                     top_logprobs=top_logprobs,
-                    num_output_top_logprobs=request.logprobs,
+                    num_output_top_logprobs=request.top_logprobs,
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/serving_chat.py:370]: There is another place where `_create_logprobs` is called with incorrect `num_output_top_logprobs`.

File: vllm/entrypoints/openai/protocol.py [hunk 2]
```diff
@@ -192,8 +192,9 @@ class ChatCompletionRequest(OpenAIBaseModel):
     # doc: end-chat-completion-extra-params
 
     def to_sampling_params(self) -> SamplingParams:
-        if self.logprobs and not self.top_logprobs:
-            raise ValueError("Top logprobs must be set when logprobs is.")
+        # We now allow logprobs being true without top_logrobs.
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/protocol.py:195]: Not necessary to comment this. `git blame` should be sufficient

**Comment by Etelis** [comment]
> > In the `_get_sampled_logprob_if_needed` function, changed `top_ranks = range(1, num_logprobs + 1)` to `top_ranks = range(1, num_logprobs)` to prevent oversampling by 1, aligning the sampling method with OpenAI's behavior.
> 
> 
> 
> Actuallly, since `range` doesn't include the end value, the original one is correct.
> 
> 
> 
> Besides, OpenAI spec indicates that up to `logprob+1` results can be returned, since the chosen token is always included even if it is not in the top-`k` results. I don't think this part is currently implemented in vLLM.
> 
> 

I must have missed that,
Thanks for noting that part.

**Comment by Etelis** [comment]
> How about we split this PR into two? We can get the first two issues fixed first before working on the third issue since that one involves more of the internals.

This sounds reasonable.
Thank you!

**Comment by Etelis** [comment]
@DarkLight1337 I have resolved the commends you left. 
Anything else you feel I need to sort?

**Comment by DarkLight1337** [comment]
Can you look into the failing tests?

**Comment by DarkLight1337** [comment]
@simon-mo To facilitate #5029 and #5031, let's merge this first.

**Comment by simon-mo** [comment]
Sounds good. Can you approve this PR and lmk when it is okay to merge?

**Review by DarkLight1337 — APPROVED** [review]

**Comment by DarkLight1337** [comment]
> Sounds good. Can you approve this PR and lmk when it is okay to merge?

It already LGTM. You can enable auto-merge.

**Comment by DarkLight1337** [comment]
@simon-mo in case you missed this

**Review by simon-mo — APPROVED** [review]

**Pull request merged**